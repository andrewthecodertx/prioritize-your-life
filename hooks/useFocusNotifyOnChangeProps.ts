import React from "react";
import { NotifyOnChangeProps } from "@tanstack/query-core";
import { useFocusEffect } from "@react-navigation/native";


/**
 * Hook that returns a function that can be used to notify the parent component
 * when the component is focused and the props have changed.
 *
 * This is useful when you want to notify the parent component when the props
 * have changed, but only when the component is focused.
 *
 * @param notifyOnChangeProps - The props to notify the parent component about
 * when they change. Can be a function that returns the props to notify about.
 */
export function useFocusNotifyOnChangeProps(
  notifyOnChangeProps?: NotifyOnChangeProps,
) {
  const focusedRef = React.useRef(true)

  useFocusEffect(
    React.useCallback(() => {
      focusedRef.current = true

      return () => {
        focusedRef.current = false
      }
    }, []),
  )

  return () => {
    if (!focusedRef.current) {
      return []
    }

    if (typeof notifyOnChangeProps === "function") {
      return notifyOnChangeProps()
    }

    return notifyOnChangeProps
  }
}
