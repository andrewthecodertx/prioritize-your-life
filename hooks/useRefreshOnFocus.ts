import React from "react";
import { useFocusEffect } from "@react-navigation/native";

/**
 * Refresh the screen when it is focused
 * @param refetch - The refetch function
 */
export function useRefreshOnFocus<T>(refetch: () => Promise<T>) {
  const firstTimeRef = React.useRef(true);

  useFocusEffect(
    React.useCallback(() => {
      if (firstTimeRef.current) {
        firstTimeRef.current = false;
        return;
      }

      refetch();
    }, [refetch])
  );
}
