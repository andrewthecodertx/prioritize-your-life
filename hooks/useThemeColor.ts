/**
 * Learn more about light and dark modes:
 * https://docs.expo.dev/guides/color-schemes/
 */

import { useColorScheme } from "react-native";
import { Colors } from "@/constants/Colors";

/**
 * Get the color for the current theme
 * @param props - The props object
 * @param colorName - The color name
 * @returns The color for the current theme
 */
export function useThemeColor(
  props: { light?: string; dark?: string },
  colorName: keyof typeof Colors.light & keyof typeof Colors.dark
) {
  const theme = useColorScheme() ?? "light";
  const colorFromProps = props[theme];

  if (colorFromProps) {
    return colorFromProps;
  } else {
    return Colors[theme][colorName];
  }
}
