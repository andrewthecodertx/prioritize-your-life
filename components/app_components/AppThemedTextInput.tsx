import React, { useState } from "react";
import {
  StyleProp,
  TextInput,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
} from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import { AppIcon } from "./AppIcon";
import { AppThemedView } from "./AppThemedView";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";

interface AppThemedTextInputProps {
  iconName?: string;
  placeholder: string;
  containerStyle?: StyleProp<ViewStyle>;
  inputStyle?: StyleProp<TextStyle>;
  secureEntry: boolean;
  value: string;
  keyboardType?: "default" | "numeric";
  checkValue: (value: string) => void;
  setValue: (value: string) => void;
}

/**
 * Represents a themed text input component.
 * @param {AppThemedTextInputProps} props - The props for the themed text input component.
 * @param {string} props.iconName - The name of the icon to display.
 * @param {string} props.placeholder - The placeholder text for the input.
 * @param {StyleProp<ViewStyle>} props.containerStyle - The custom style for the container.
 * @param {StyleProp<TextStyle>} props.inputStyle - The custom style for the input.
 * @param {boolean} props.secureEntry - Indicates if the input is a password field.
 * @param {string} props.value - The value of the input.
 * @param {(value: string) => void} props.checkValue - The function to check the value of the input.
 * @param {(value: string) => void} props.setValue - The function to set the value of the input.
 * @returns {JSX.Element} The themed text input component.
 */
export const AppThemedTextInput = ({
  checkValue,
  iconName,
  placeholder,
  secureEntry,
  setValue,
  containerStyle,
  inputStyle,
  value,
  keyboardType,
  ...otherProps
}: AppThemedTextInputProps) => {
  const textColor = useThemeColor({}, "text");
  let backgroundColor = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );

  const [isPasswordVisible, setIsPasswordVisible] = useState(!secureEntry);

  return (
    <AppThemedView style={[styles.inputContainer, containerStyle]}>
      <TextInput
        keyboardType={keyboardType}
        onBlur={(e) => {
          if (value.length > 0) checkValue(value);
        }}
        onChangeText={(text) => setValue(text)}
        placeholder={placeholder}
        placeholderTextColor="#999"
        secureTextEntry={!isPasswordVisible}
        style={[
          styles.input,
          { backgroundColor: backgroundColor, color: textColor },
          inputStyle,
        ]}
        value={value}
        {...otherProps}
      />
      {secureEntry && (
        <TouchableOpacity
          onPress={() => setIsPasswordVisible(!isPasswordVisible)}
          testID="passwordVisibilityToggle"
        >
          <AppIcon
            name={!isPasswordVisible ? "eye-off" : "eye"}
            size={s(24)}
            color="#ccc"
          />
        </TouchableOpacity>
      )}

      {iconName && <AppIcon name={iconName} size={s(24)} color="#ccc" />}
    </AppThemedView>
  );
};

const styles = ScaledSheet.create({
  inputContainer: {
    flexDirection: "row",
    alignItems: "center",
    borderColor: "#ccc",
    borderWidth: s(1),
    borderRadius: s(5),
    marginBottom: s(10),
    paddingHorizontal: s(10),
    width: "80%",
  },
  input: {
    flex: 1,
    padding: s(10),
  },
});

export default AppThemedTextInput;
