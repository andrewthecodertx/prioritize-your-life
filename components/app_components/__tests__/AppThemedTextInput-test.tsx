import React from "react";
import { render, fireEvent, screen } from "@testing-library/react-native";
import AppThemedTextInput from "../AppThemedTextInput";
import { useThemeColor } from "@/hooks/useThemeColor";

jest.mock("@/hooks/useThemeColor", () => ({
  useThemeColor: jest.fn(),
}));

jest.mock("../AppIcon", () => ({
  AppIcon: jest.fn(() => null),
}));

describe("AppThemedTextInput", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const mockTextColor = "#000";
  const mockBackgroundColor = "#fff";

  beforeEach(() => {
    (useThemeColor as jest.Mock).mockImplementation((colors, key) => {
      return key === "text" ? mockTextColor : mockBackgroundColor;
    });
  });

  it("renders correctly and matches snapshot", () => {
    const { toJSON } = render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value=""
        setValue={() => {}}
        checkValue={() => {}}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("displays the placeholder text", () => {
    render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value=""
        setValue={() => {}}
        checkValue={() => {}}
      />
    );
    expect(screen.getByPlaceholderText("Enter text")).toBeTruthy();
  });

  it("handles text input changes", () => {
    const mockSetValue = jest.fn();
    render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value=""
        setValue={mockSetValue}
        checkValue={() => {}}
      />
    );
    fireEvent.changeText(screen.getByPlaceholderText("Enter text"), "New text");
    expect(mockSetValue).toHaveBeenCalledWith("New text");
  });

  it("calls checkValue on blur if value is not empty", () => {
    const mockCheckValue = jest.fn();
    render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value="Some value"
        setValue={() => {}}
        checkValue={mockCheckValue}
      />
    );
    fireEvent(screen.getByPlaceholderText("Enter text"), "blur");
    expect(mockCheckValue).toHaveBeenCalledWith("Some value");
  });

  // it("toggles password visibility when eye icon is pressed", () => {
  // #TODO - Add test for this
  // });

  // it("renders an icon if iconName prop is provided", () => {
  // #TODO - Add test for this
  // });
});
