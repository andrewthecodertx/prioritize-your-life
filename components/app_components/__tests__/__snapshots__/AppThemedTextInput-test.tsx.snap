// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemedTextInput Tests renders correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderColor": "#ccc",
        "borderRadius": 10.714285714285714,
        "borderWidth": 2.142857142857143,
        "flexDirection": "row",
        "marginBottom": 21.428571428571427,
        "paddingHorizontal": 21.428571428571427,
        "width": "80%",
      },
      undefined,
    ]
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Snapshot test!"
    placeholderTextColor="#999"
    secureTextEntry={false}
    style={
      [
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
        {
          "backgroundColor": "#151718",
          "color": undefined,
        },
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
      ]
    }
    value=""
  />
</View>
`;

exports[`AppThemedTextInput Tests renders correctly with dark theme 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderColor": "#ccc",
        "borderRadius": 10.714285714285714,
        "borderWidth": 2.142857142857143,
        "flexDirection": "row",
        "marginBottom": 21.428571428571427,
        "paddingHorizontal": 21.428571428571427,
        "width": "80%",
      },
      undefined,
    ]
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Snapshot test!"
    placeholderTextColor="#999"
    secureTextEntry={false}
    style={
      [
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
        {
          "backgroundColor": "#000000",
          "color": "#000000",
        },
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
      ]
    }
    value=""
  />
</View>
`;

exports[`AppThemedTextInput Tests renders correctly with light theme 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderColor": "#ccc",
        "borderRadius": 10.714285714285714,
        "borderWidth": 2.142857142857143,
        "flexDirection": "row",
        "marginBottom": 21.428571428571427,
        "paddingHorizontal": 21.428571428571427,
        "width": "80%",
      },
      undefined,
    ]
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Snapshot test!"
    placeholderTextColor="#999"
    secureTextEntry={false}
    style={
      [
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
        {
          "backgroundColor": "#ffffff",
          "color": "#ffffff",
        },
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
      ]
    }
    value=""
  />
</View>
`;
